src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/ProxyHelper.java

```diff
@@ -174,23 +209,81 @@ public static Proxy createProxy(@Nullable String proxyAddress) throws IOExceptio
+
+    // Determine credentials: URL credentials take precedence over system properties
+    String username = urlUsername;
+    String password = urlPassword;
+
     if (username != null) {
       if (password == null) {
         throw new IOException("No password given for proxy " + cleanProxyAddress);
       }
-
-      // We need to make sure the proxy password is not url encoded; some special characters in
+      // We need to make sure the proxy credentials are not url encoded; some special characters in
       // proxy passwords require url encoding for shells and other tools to properly consume.
-      final String decodedPassword = URLDecoder.decode(password, "UTF-8");
-      Authenticator.setDefault(
-          new Authenticator() {
-            @Override
-            public PasswordAuthentication getPasswordAuthentication() {
-              return new PasswordAuthentication(username, decodedPassword.toCharArray());
-            }
-          });
+      username = URLDecoder.decode(username, "UTF-8");
+      password = URLDecoder.decode(password, "UTF-8");
+    } else if (systemPropertyUser != null && systemPropertyPassword != null) {
+      // Fall back to system property credentials
+      username = systemPropertyUser;
+      password = systemPropertyPassword;
     }
 
-    return new Proxy(Proxy.Type.HTTP, new InetSocketAddress(hostname, port));
+    // If credentials are provided, also set up Java's Authenticator for HTTPS proxy support.
+    // For HTTPS connections through HTTP proxies (CONNECT tunneling), Java's HttpURLConnection
+    // handles the CONNECT request internally and won't use Proxy-Authorization header we set.
+    // Instead, it uses the Authenticator mechanism. We also enable Basic auth tunneling by
+    // clearing the disabled schemes (by default, Basic auth is disabled for HTTPS tunneling).
+    if (username != null && password != null) {
+      // Use double-checked locking to ensure thread-safe, one-time setup of the global
+      // Authenticator. The first caller with credentials wins. This is safe because Bazel
+      // typically uses a single proxy configuration for all downloads.
+      if (!authenticatorSet) {
+        synchronized (AUTHENTICATOR_LOCK) {
+          if (!authenticatorSet) {
+            final String finalUsername = username;
+            final String finalPassword = password;
+            Authenticator.setDefault(
+                new Authenticator() {
+                  @Override
+                  public PasswordAuthentication getPasswordAuthentication() {
+                    // Only provide credentials for proxy authentication, not for other auth
+                    // challenges. This prevents credential leakage to non-proxy contexts.
+                    if (getRequestorType() == RequestorType.PROXY) {
+                      return new PasswordAuthentication(finalUsername, finalPassword.toCharArray());
+                    }
+                    return null;
```

Comment:
Doesn't this break existing usage without a proxy but with authorized access by no longer returning credentials in that case?